arm: put saving of FPU context into its own file so it can be reused

oyvindronningstad opened the pull request:
Also, to ease readability.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

oyvindronningstad committed:
tfm: Put saving of FPU context into its own file so it can be reused

Also, this eases readability.

The new API can be used any time all FP registers must be manually
saved and restored for an operation.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

ioannisg reviewed (changes requested):
Assuming this is a "no-op", I am ok with that.

I would like, though the added functions to be provided in the internal arch/arm api, because they seem to have little to do with TF-M.

Would that be possible @oyvindronningstad ? E.g as part of thread.c or scb.c in the AARC32 architecture? You could leave the FPU_SHARING in the implementation, but I suppose you could keep the #if defined(CONFIG_ARM_NONSECURE_PREEMPTIBLE_SECURE_CALLS) inside the caller, that is: inside interface.c. Cause the functions could have a more generic use...

oyvindronningstad commented:
@ioannisg  thanks, moved to arch/arm/aarch32/cortex_m like discussed offline. Moved the #ifdef CONFIG_ARM_NONSECURE_PREEMPTIBLE_SECURE_CALLS back to interface.c

ioannisg commented on arch/arm/core/aarch32/cortex_m/fpu.c:
```diff
@@ -0,0 +1,74 @@
+#include <arch/arm/aarch32/cortex_m/cmsis.h>
+
+/**
+ * @file @brief Helper functions for save and restoring the FP context
```
nit: for saving and restoring the FP context

ioannisg commented on arch/arm/core/aarch32/cortex_m/fpu.c:
```diff
@@ -0,0 +1,74 @@
+ *
+ */
+
+int save_fp_context(uint32_t *buffer, size_t buffer_size_words, bool *context_saved)
```
Since the functions are now part of the ARCH/ARM internal API, then need to have the `z_arm_` prefix, as most of the exposed internal API

ioannisg commented on arch/arm/core/aarch32/cortex_m/fpu.c:
```diff
@@ -0,0 +1,74 @@
+int save_fp_context(uint32_t *buffer, size_t buffer_size_words, bool *context_saved)
+{
+#if defined(CONFIG_FPU_SHARING)
+	if (buffer_size_words < 33) {
```
Could the 33 be a macro, for ease of understanding, or, alternatively, add a comment?

ioannisg reviewed (approved):
Thanks @oyvindronningstad 
I left a few comments for you to address.

ioannisg commented on modules/trusted-firmware-m/interface/interface.c:
```diff
@@ -45,53 +49,18 @@ int32_t tfm_ns_interface_dispatch(veneer_fn fn,
-	uint32_t fp_ctx_callee_saved[16];
-	uint32_t fp_ctx_FPSCR;
+	uint32_t context_buffer[33];
 	bool context_saved = false;
```
@oyvindronningstad i think this variable needs to be inside the #ifdef as well, to suppress the unused variable error

stephanosio commented on modules/trusted-firmware-m/interface/interface.c:
```diff
@@ -24,6 +24,10 @@
 K_MUTEX_DEFINE(tfm_mutex);
 
+
+int z_arm_save_fp_context(uint32_t *buffer, size_t buffer_size_words, bool *context_saved);
+void z_arm_restore_fp_context(uint32_t *buffer, bool context_saved);
```
If reusability is the goal here, these should probably go somewhere into `include/arch/arm/aarch32/` (maybe `exc.h`?).

Also, it would be nice to refactor "FP context" struct into its own and use it here:
https://github.com/zephyrproject-rtos/zephyr/blob/21d1ad3762302b3e461953df59430c77e0709274/include/arch/arm/aarch32/exc.h#L97-L98

ioannisg replied:
Right. I was thinking that we could decide on adding the declarations, once we actually have the re-use case in some "in-tree" use-case. So I felt it is ok for now to keep the declarations where the functions are used. But sure, we could declare of course the functions in this arch file

oyvindronningstad replied:
>Also, it would be nice to refactor "FP context" struct into its own and use it here:

The context is not quite the same. In exc.h, only 16 of the 32 fp registers are in the struct (the callee saved registers). But I can make an analogous struct.

stephanosio commented on modules/trusted-firmware-m/interface/interface.c:
```diff
@@ -45,53 +49,18 @@ int32_t tfm_ns_interface_dispatch(veneer_fn fn,
-	}
-#endif /* defined(CONFIG_ARM_NONSECURE_PREEMPTIBLE_SECURE_CALLS) && defined(CONFIG_FPU_SHARING) */
+
+	z_arm_save_fp_context(context_buffer, sizeof(context_buffer) / 4, &context_saved);
```
Should use `ARRAY_SIZE` here

stephanosio commented on arch/arm/core/aarch32/cortex_m/fpu.c:
```diff
@@ -0,0 +1,78 @@
+	return 0;
+}
+
+void z_arm_restore_fp_context(uint32_t *buffer, bool context_saved)
```
would be nice to add `const` here.

oyvindronningstad replied:
done

stephanosio commented on arch/arm/core/aarch32/cortex_m/fpu.c:
```diff
@@ -0,0 +1,78 @@
+#define FPSCR_OFFS 32
+#define BUF_SIZE (FPSCR_OFFS + 1)
+
+int z_arm_save_fp_context(uint32_t *buffer, size_t buffer_size_words, bool *context_saved)
```
I wonder if it would be better to define a struct for "FP context" and use that here instead. Also, it would be nice to have some integration with the `__esf` struct, in that regard:

https://github.com/zephyrproject-rtos/zephyr/blob/21d1ad3762302b3e461953df59430c77e0709274/include/arch/arm/aarch32/exc.h#L85

ioannisg replied:
Of it includes callee saved, not sure if it makes a lot of sense integrate with the exception stack frame...

oyvindronningstad replied:
Made a stuct

stephanosio reviewed (commented):
Just some suggestions

oyvindronningstad committed:
tests: arm_thread_swap_tz: Enhance testing of FP regs

The existing test puts the same value into all FP registers, runs a
secure service which is then interrupted by a thread that clobbers the
registers.

This patch instead puts different values into each register to test
that each register is restored in the right order.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

yashi commented:
@oyvindronningstad, it'd be nice if you could work on the commit messages a bit.  The subject line should be a summary of the body.  And the start of the body shouldn't be continuous from the subject.  It's always a good idea to write the body first and then summarize your message and put it in the subject line at the last.

https://docs.zephyrproject.org/latest/contribute/index.html#commit-guidelines

oyvindronningstad commented:
@yashi I updated now, please have a look.

yashi commented:
@oyvindronningstad, thanks.  That's much better. :smiley:

I just noticed that your commit is actually a bug fix rather than moving a function for re-usability nor readability, no?  Why don't you create a commit before or after moving the code saying, "arm: Fix restore order for FPU regs" or something along the the line; so that your test case can be "arm: Test restore order for FPU regs" to go with it.

Hope I'm not asking too much.

ioannisg commented:
> I just noticed that your commit is actually a bug fix rather than moving a function for re-usability nor readability, no? Why don't you create a commit before or after moving the code saying, "arm: Fix restore order for FPU regs" or something along the the line; so that your test case can be "arm: Test restore order for FPU regs" to go with it.

If this is true (which it seems it is, since the vldmia is called with the same argument, AFAICS) I agree this should have the bug label.

@oyvindronningstad the nrf git tag is not needed here ;)

stephanosio commented on include/arch/arm/aarch32/cortex_m/fpu.h:
```diff
@@ -0,0 +1,20 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _CORTEX_M_FPU_H
```
Names starting with `_` is deprecated.
```suggestion
#ifndef ZEPHYR_INCLUDE_ARCH_ARM_AARCH32_CORTEX_M_FPU_H_
```

ioannisg replied:
@oyvindronningstad ^^

oyvindronningstad replied:
done

stephanosio reviewed (commented):
Looks ok overall.

oyvindronningstad commented:
The bug fix has been moved to https://github.com/zephyrproject-rtos/zephyr/pull/35954

ioannisg commented:
@oyvindronningstad can you rebase this PR now, to exclude the hotfix and keep only the refactoring work? (you'll need to work the commit messages, too :) )

oyvindronningstad commented:
@ioannisg Thanks, done
